### PR TITLE
refactor: swap order of partition and table in in-memory catalog

### PR DESCRIPTION
**Pull request by tustvold**
This PR swaps the order of partitions and tables within the in-memory catalog. This will be the first in a number of PRs to simplify the in-memory catalog and bring it closer to @pauldix 's vision (which i share) of a less concurrent but simpler catalog and accompanying lifecycle.

This PR looks a lot longer than it is, because it swaps the order of partition key and table name in a number of methods

### CODE REVIEW

File: data_types/src/partition_metadata.rs [hunk 1]
```diff
@@ -28,57 +27,36 @@ impl PartitionSummary {
...
-    pub key: String,
-
-    /// The chunks of the tables in this partition
-    pub tables: Vec<UnaggregatedTableSummary>,
 }
 
 /// Metadata and statistics for a Chunk *within* a partition
 #[derive(Debug, Deserialize, Serialize, PartialEq, Clone)]
-pub struct UnaggregatedTableSummary {
+pub struct PartitionChunkSummary {
```
Comment (tustvold) [inline data_types/src/partition_metadata.rs:39]: I renamed this to what I think is a slightly more obvious name.

File: server/src/db.rs [hunk 2]
```diff
@@ -104,41 +96,6 @@ pub enum Error {
         action: String,
     },
 
-    #[snafu(display(
```
Comment (tustvold) [inline server/src/db.rs:107]: Using identical error variants with the name of the method seems a tad redundant - the caller should already know what method the error was returned from. Simply propagating the error from the catalog if any eliminates a lot of error plumbing

File: server/src/db.rs [hunk 3]
```diff
@@ -1051,20 +948,23 @@ impl Db {
...
-                let partition = self
-                    .preserved_catalog
-                    .state()
-                    .get_or_create_partition(partition_key);
-                let mut partition = partition.write();
-                partition.update_last_write_at();
-
                 for table_batch in write.table_batches() {
-                    match partition.open_chunk(table_batch.name()).ok().flatten() {
+                    if table_batch.row_count() == 0 {
```
Comment (tustvold) [inline server/src/db.rs:955]: Avoid potentially creating an empty open chunk

File: server/src/db.rs [hunk 4]
```diff
@@ -2875,11 +2735,12 @@ mod tests {
             .with_labels(&[
                 ("db_name", "lock_tracker"),
                 ("lock", "partition"),
+                ("table", "cpu"),
                 ("svr_id", "10"),
                 ("access", "shared"),
             ])
             .counter()
-            .eq(2.)
+            .eq(1.)
```
Comment (tustvold) [inline server/src/db.rs:2743]: No longer acquire a lock to get the partition key

File: server/src/db/catalog.rs [hunk 5]
```diff
@@ -1,129 +1,49 @@
...
-    #[snafu(display(
-        "Internal Error: unexpected chunk state for {}:{}:{}  during {}. Expected {}, got {}",
-        partition_key,
-        table_name,
-        chunk_id,
-        operation,
-        expected,
-        actual
-    ))]
-    InternalChunkState {
```
Comment (tustvold) [inline server/src/db/catalog.rs:52]: The lifecycle related variants are moved into a separate enumeration for the catalog chunk

File: server/src/db/catalog.rs [hunk 6]
```diff
@@ -187,102 +109,142 @@ impl Catalog {
...
+    pub fn partitions(&self) -> Vec<Arc<RwLock<Partition>>> {
+        self.tables
+            .read()
+            .values()
+            .flat_map(|table| table.partitions().cloned())
+            .collect()
+    }
+
+    /// Get a specific table by name, returning `None` if there is no such table
+    pub fn table(&self, table_name: impl AsRef<str>) -> Result<MappedRwLockReadGuard<'_, Table>> {
```
Comment (tustvold) [inline server/src/db/catalog.rs:129]: This MappedRwLockReadGuard shenanigans will likely be removed as part of the locking rework

File: server/src/db/catalog/partition.rs [hunk 7]
```diff
@@ -87,197 +94,89 @@ impl Partition {
...
 
     /// Return a iterator over chunks in this partition
     pub fn chunks(&self) -> impl Iterator<Item = &Arc<RwLock<Chunk>>> {
-        self.tables.values().flat_map(|table| table.chunks.values())
-    }
-
-    /// Return an iterator over chunks in this partition
-    ///
-    /// `table_names` specifies which tables to include
-    pub fn filtered_chunks<'a>(
```
Comment (tustvold) [inline server/src/db/catalog/partition.rs:236]: Partitions no longer contain multiple tables so this is unnecessary

File: server/src/db/system_tables.rs [hunk 8]
```diff
@@ -934,15 +857,12 @@ mod tests {
             "+---------------+----------+------------+-------------+-------------------+-------+-----------+-----------+-----------------+",
             "| p1            | 42       | t1         | c1          | ReadBuffer        | 55    | bar       | foo       | 11              |",
             "| p1            | 42       | t1         | c2          | ReadBuffer        | 66    | 11        | 43        | 12              |",
-            "| p1            | 42       | t1         | __other     | ReadBuffer        |       |           |           | 13              |",
-            "| p2            | 43       | t1         | c2          | OpenMutableBuffer | 667   | 110       | 430       |                 |",
-            "| p2            | 43       | t1         | c1          | OpenMutableBuffer |       |           |           | 100             |",
+            "| p2            | 43       | t1         | c1          | OpenMutableBuffer | 667   | 110       | 430       | 100             |",
```
Comment (tustvold) [inline server/src/db/system_tables.rs:860]: The previous code handled the case when the TableSummary and DetailedChunkSummary had different columns. As these are now collected simultaneously for a given chunk I don't think this additional complexity is necessary. @alamb ?

File: tracker/src/lock.rs [hunk 9]
```diff
@@ -1,11 +1,16 @@
...
 use metrics::{KeyValue, MetricObserver, MetricObserverBuilder};
 use std::sync::atomic::{AtomicU64, Ordering};
 use std::sync::Arc;
 use std::time::Duration;
 
+type RawRwLock = InstrumentRawRwLock<parking_lot::RawRwLock>;
+
 /// An instrumented Read-Write Lock
-pub type RwLock<T> = lock_api::RwLock<InstrumentRawRwLock<parking_lot::RawRwLock>, T>;
+pub type RwLock<T> = lock_api::RwLock<RawRwLock, T>;
```
Comment (tustvold) [inline tracker/src/lock.rs:9]: The API for these favours being explicit over concise, and to use certain methods you need to specify the full type.

E.g

```
pub fn table(&self, table_name: impl AsRef<str>) -> Result<MappedRwLockReadGuard<'_, Table>> {
        let table_name = table_name.as_ref();
        RwLockReadGuard::try_map(self.tables.read(), |tables| tables.get(table_name)).map_err(
            |_| Error::TableNotFound {
                table: table_name.to_string(),
            },
        )
    }
```

File: server/src/db/catalog.rs [hunk 10]
```diff
@@ -187,102 +109,142 @@ impl Catalog {
...
-        self.filtered_chunks(partition_key, table_names, Chunk::summary)
+        self.filtered_chunks(table_names, partition_key, Chunk::summary)
     }
 
-    pub fn detailed_chunk_summaries(&self) -> Vec<DetailedChunkSummary> {
+    pub fn detailed_chunk_summaries(&self) -> Vec<(Arc<TableSummary>, DetailedChunkSummary)> {
         let partition_key = None;
         let table_names = TableNameFilter::AllTables;
-        self.filtered_chunks(partition_key, table_names, Chunk::detailed_summary)
+        // TODO: Having two summaries with overlapping information seems unfortunate
```
Comment (tustvold) [inline server/src/db/catalog.rs:232]: I think it would be good to be able to create a TableSummary from a DetailedChunkSummary and therefore avoid this duplication

File: server/src/db/catalog/partition.rs [hunk 11]
```diff
@@ -87,197 +94,89 @@ impl Partition {
...
 
     /// Create new chunk that is only in object store (= parquet file).
     ///
     /// The table-specific chunk ID counter will be set to
     /// `max(current, chunk_id + 1)`.
     ///
-    /// Fails if the chunk already exists.
-    pub fn create_object_store_only_chunk(
+    /// Returns the previous chunk with the given chunk_id if any
+    pub fn insert_object_store_only_chunk(
```
Comment (tustvold) [inline server/src/db/catalog/partition.rs:128]: The semantics of this were previously a bit odd. If the chunk with the same id already existed it would replace it and return an error. I've made it follow the semantics of a typical insert method.

I also updated the callsite to just panic if it gets a collision as there isn't an obvious way to recover

FYI @crepererum

File: data_types/src/partition_metadata.rs [hunk 12]
```diff
@@ -5,18 +5,17 @@ use std::{borrow::Cow, cmp::Ordering, mem};
 
 use serde::{Deserialize, Serialize};
 use std::borrow::Borrow;
+use std::iter::FromIterator;
 use std::num::NonZeroU64;
 
 /// Describes the aggregated (across all chunks) summary
-/// statistics for each column in each table in a partition
+/// statistics for each column in a partition
```
Comment (tustvold) [inline data_types/src/partition_metadata.rs:12]: Further simplification of these metadata structs is likely possible

File: server/src/db/catalog/metrics.rs [hunk 13]
```diff
@@ -35,22 +25,53 @@ impl CatalogMetrics {
         &self.memory_metrics
     }
 
-    pub(super) fn new_lock<T>(&self, t: T) -> RwLock<T> {
-        self.lock_tracker.new_lock(t)
-    }
+    pub(super) fn new_table_metrics(&self, table_name: &str) -> TableMetrics {
```
Comment (tustvold) [inline server/src/db/catalog/metrics.rs:28]: This changes the lock metrics from being collected per-partition to per-table. I'm not sure if this will cause a cardinality explosion that will make our clusters unhappy. @e-dard ?

File: server/src/db/catalog/partition.rs (hunk 11 above)
Comment (crepererum) [reply server/src/db/catalog/partition.rs:128]: I think overwriting an existing chunk with a new open one should be an error. Otherwise:

- the whole state machine is somewhat broken (you can "go back" to an open chunk)
- the preserved catalog can break (it can reference chunk IDs twice)
- having unique chunk IDs is a very fundamental assumption all over the place

Comment (crepererum) [reply server/src/db/catalog/partition.rs:128]: technically I think you can just panic, because if this ever happens then some lifecycle invariant is violated.

Comment (tustvold) [reply server/src/db/catalog/partition.rs:128]: So would you like me to make this method panic, instead of the callsite?

Comment (crepererum) [reply server/src/db/catalog/partition.rs:128]: I would suggest a combination of the following:

1. panic here aka "locally" (because the callsite might change / be refactored at some point and people might forget about it)
2. don't replace the existing entry (aka "panic before modify")

File: data_types/src/partition_metadata.rs [hunk 14]
```diff
@@ -5,18 +5,17 @@ use std::{borrow::Cow, cmp::Ordering, mem};
...
 #[derive(Debug, Deserialize, Serialize, PartialEq)]
 pub struct PartitionSummary {
     /// The identifier for the partition, the partition key computed from
     /// PartitionRules
     pub key: String,
-
-    /// The tables in this partition
-    pub tables: Vec<TableSummary>,
+    pub table: TableSummary,
```
Comment (alamb) [inline data_types/src/partition_metadata.rs:18]: 👍

File: data_types/src/partition_metadata.rs [hunk 15]
```diff
@@ -28,57 +27,36 @@ impl PartitionSummary {
...
-pub struct UnaggregatedTableSummary {
+pub struct PartitionChunkSummary {
     pub chunk_id: u32,
     pub table: TableSummary,
 }
 
+impl FromIterator<TableSummary> for TableSummary {
+    fn from_iter<T: IntoIterator<Item = TableSummary>>(iter: T) -> Self {
+        let mut iter = iter.into_iter();
+        let first = iter.next().expect("must contain one element");
```
Comment (alamb) [inline data_types/src/partition_metadata.rs:47]:
```suggestion
        let first = iter.next().expect("must contain one least element");
```

File: influxdb_iox_client/src/client/management.rs [hunk 16]
```diff
@@ -561,12 +561,12 @@ impl Client {
...
     }
 
-    /// Create a new chunk in a partittion
+    /// Create a new chunk in a partition
     pub async fn new_partition_chunk(
         &mut self,
         db_name: impl Into<String> + Send,
-        partition_key: impl Into<String> + Send,
         table_name: impl Into<String> + Send,
+        partition_key: impl Into<String> + Send,
```
Comment (alamb) [inline influxdb_iox_client/src/client/management.rs:569]: This kind of change is always a little tricky -- swap the meaning of two parameters but the types are the same (so the compiler can't help you); Just observing

File: server/src/db/catalog/chunk.rs [hunk 17]
```diff
@@ -230,15 +293,8 @@ impl Chunk {
...
-    ) -> Result<Self> {
-        ensure!(
-            chunk.rows() > 0,
-            ChunkIsEmpty {
-                partition_key: partition_key.as_ref(),
-                table_name: chunk.table_name().as_ref(),
-                chunk_id,
-            }
-        );
+    ) -> Self {
```
Comment (alamb) [inline server/src/db/catalog/chunk.rs:296]: why the change to `panic` from fallable function?

File: server/src/db/catalog/metrics.rs (hunk 13 above)
Comment (alamb) [reply server/src/db/catalog/metrics.rs:28]: It is actually "per-partition, per table" right?

File: server/src/db.rs (hunk 2 above)
Comment (alamb) [reply server/src/db.rs:107]: the downside is that you then don't have the "callchain" of what the path was.

However, I also find it quite annoying to have to (effectively) build up manual stack traces.

File: server/src/db.rs [hunk 18]
```diff
@@ -448,30 +399,20 @@ impl Db {
...
-            .state()
-            .valid_partition(partition_key)
-            .context(RollingOverPartition {
-                partition_key,
-                table_name,
-            })?;
+        let chunk = self
+            .partition(table_name, partition_key)?
+            .read()
+            .open_chunk();
```
Comment (alamb) [inline server/src/db.rs:408]: So now this error will say only "catalog error" rather than "catalog error while rolling over partition"

That seems alike a regression in usefulness to me, though I don't feel strongly enough about it to insist on it being fixed. In general, it seems like if someone has gone through the effort to put in more error detail, it would be nice to not remove it subsequently

File: server/src/db.rs (hunk 3 above)
Comment (alamb) [reply server/src/db.rs:955]: Ah, here is why you made new chunk panic!

File: server/src/db.rs [hunk 19]
```diff
@@ -2773,33 +2667,6 @@ mod tests {
 
         let db = Arc::new(test_db.db);
 
-        // Expect partition lock registry to have been created
-        test_db
```
Comment (alamb) [inline server/src/db.rs:2777]: Should these tests be updated rather than removed?

File: server/src/db/system_tables.rs (hunk 8 above)
Comment (alamb) [reply server/src/db/system_tables.rs:860]: The usecase as I recall for an extra column in `DetailedChunkSummary` was a synthetic `__dictionary` column to report on the size of the (shared) dictionary across all tables. As I believe now each column has its own dictionary this isn't needed anymore

File: server/src/db/catalog.rs (hunk 10 above)
Comment (alamb) [reply server/src/db/catalog.rs:232]: Sounds like a good improvement to me

**Review by alamb — APPROVED** [review]
I can't say I was able to grok every line of this PR, but I reviewed it carefully and it looks like a solid improvement to me -- I had some minor suggestions / comments but nothing that would prevent this from merging. Nice work @tustvold 

I also think this PR should be merged sooner rather than later as it will conflict with / cause conflict with pretty much anything else ;)

File: server/src/db/catalog/chunk.rs (hunk 17 above)
Comment (tustvold) [reply server/src/db/catalog/chunk.rs:296]: Because it seems unfortunate to add error propagation logic for a situation that should never realistically occur. The error pattern we have in IOx has a tendency to result in large Error enumerations with a huge number of variants, which then the callers are burdened with correctly handling. I dunno, I find the amount of time spent shuffling errors around the place incredibly frustrating, not to mention distracting

File: server/src/db/catalog/metrics.rs (hunk 13 above)
Comment (tustvold) [reply server/src/db/catalog/metrics.rs:28]: Nope, just per-table. A single value will be computed across all partitions within a given table.

File: server/src/db.rs (hunk 18 above)
Comment (tustvold) [reply server/src/db.rs:408]: I'd argue it isn't more informative to get the error `catalog error: partition not found` vs `catalog error while rolling over partition: partition not found` whilst calling a method called `rollover_partition`?

If rollover_partition was being called by some other method that say rolled over a partition and then did some other fallible thing, the burden should be on this method to add context to distinguish these errors by adding further context? Pushing all context down to the lowest level in the error hierarchy seems to not only break encapsulation, but also seems to be unnecessary?

File: server/src/db.rs (hunk 19 above)
Comment (tustvold) [reply server/src/db.rs:2777]: The metric now won't be reported until there has been a write and consequently one or more tables exist

File: server/src/db/catalog/chunk.rs (hunk 17 above)
Comment (crepererum) [reply server/src/db/catalog/chunk.rs:296]: Technically this error should never occur since the lifecycle/write logic should never create an empty chunk. So it's more of a sanity/invariant check and hitting this is a hard bug somewhere in our code.

File: server/src/db/catalog/chunk.rs [hunk 20]
```diff
@@ -802,23 +859,16 @@ mod tests {
...
             mb_chunk,
             ChunkMetrics::new_unregistered(),
-        )
-        .unwrap();
+        );
         assert!(matches!(chunk.stage(), &ChunkStage::Open { .. }));
+    }
 
+    #[test]
+    #[should_panic]
```
Comment (crepererum) [inline server/src/db/catalog/chunk.rs:867]:
```suggestion
    #[should_panic(expected = "chunk must not be empty")]
```

Personally I think we should always also check for the panic message so it really hits the error case we intended and not something else (like division by zero).

File: server/src/db/catalog/table.rs [hunk 21]
```diff
@@ -0,0 +1,72 @@
+use super::partition::Partition;
```
Comment (crepererum) [inline server/src/db/catalog/table.rs:1]: Extra points for further splitting up the catalog code :+1:

**Review by crepererum — APPROVED** [review]
Only minor nitpick, looks good overall (although I have to admit that re-shuffle changes are naturally hard to review).

File: server/src/db/catalog/chunk.rs (hunk 20 above)
Comment (tustvold) [reply server/src/db/catalog/chunk.rs:867]: I didn't realise you could do this - nice :ok_hand:

File: server/src/db.rs (hunk 18 above)
Comment (alamb) [reply server/src/db.rs:408]: We talked face to face and I think this PR is ready to merge as is 👍
